src/Orleans.Core/Diagnostics/ActivityPropagationGrainCallFilter.cs

```diff
@@ -0,0 +1,94 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Orleans.Runtime
+{
+    internal static class ActivityPropagationGrainCallFilter
+    {
+        private const string TraceParentHeaderName = "traceparent";
+        private const string TraceStateHeaderName = "tracestate";
+
+        internal const string ActivitySourceName = "orleans.runtime.graincall";
+        internal const string ActivityNameIn = "Orleans.Runtime.GrainCall.In";
+        internal const string ActivityNameOut = "Orleans.Runtime.GrainCall.Out";
+
+        private static readonly ActivitySource activitySource = new(ActivitySourceName);
+
+        private static async Task ProcessNewActivity(IGrainCallContext context, string activityName, ActivityKind activityKind, ActivityContext activityContext)
+        {
+            // rpc attributes from https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/rpc.md
+            ActivityTagsCollection tags = null;
+            if (activitySource.HasListeners())
+                tags = new ActivityTagsCollection
+                {
+                    {"rpc.service", context.InterfaceMethod?.DeclaringType},
+                    {"rpc.method", context.InterfaceMethod?.Name},
+                    {"net.peer.name", context.Grain},
```

Comment:
To clarify:

the opentelemetry spec requires to set the `Status` of a span, see
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/rpc.md#grpc-status
and
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#set-status

To set this, you can use the `Activity.SetStatus` method, see:
https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.activity.setstatus?view=net-6.0&viewFallbackFrom=net-5.0

Reply:
Also: I think we should add the 'peer.service' tag, and include the local siloname (available from `ILocalSiloDetails` service)
the spec says:
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/span-general.md#general-remote-service-attributes.

so `rpc.service` should contain the `Grain` interface and `peer.service` the name of the silo.

and maybe, we should also include the `rpc.system` tag with a fixed value of `"orleans"`?

Reply:
> To set this, you can use the `Activity.SetStatus` method, see:
> https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.activity.setstatus?view=net-6.0&viewFallbackFrom=net-5.0

Hello, @rikbosch could you please clarify how to set status of activity in .net 5. SetStatus exists only in .net 6.

Reply:
> Also: I think we should add the 'peer.service' tag, and include the local siloname (available from `ILocalSiloDetails` service)
> so `rpc.service` should contain the `Grain` interface and `peer.service` the name of the silo.

Is it ok to set dynamic name of the silo here? In example it has some user-friendly name.

Reply:
> > To set this, you can use the `Activity.SetStatus` method, see:
> > https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.activity.setstatus?view=net-6.0&viewFallbackFrom=net-5.0
> 
> Hello, @rikbosch could you please clarify how to set status of activity in .net 5. SetStatus exists only in .net 6.

Hi @Ilchert, I think you could set a tag on the activity with the name "status" and (constant) values of "Ok" or "Error".

> Is it ok to set dynamic name of the silo here? In example it has some user-friendly name.

The documentation does not specify any restrictions if I read it correctly:

peer.service | string | Remote service name (for some unspecified definition of "service"). E.g., "elasticsearch", "a_custom_microservice", "memcache"
-- | -- | --

Reply:
@rikbosch I added `status` tag, but I am afraid, that `peer.service` cannot be applicable to orleans.  
Because it is initial support of distributed tracing, I think we will have a several PR with tracing improvements later with a lot of additional tracing attributes.
 
Can I mark this conversation as resolved?

Reply:
@Ilchert As far as I'm concerned this can be resolved, custom tags can always be added by a custom filter further up the pipeline.

Reply:
@rikbosch, thanks for comments, I was glad to work with you under this PR.